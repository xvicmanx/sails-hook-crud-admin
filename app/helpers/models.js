import React from 'react';
const template = require('lodash.template');
import { Popup, Button } from 'semantic-ui-react';
import JSONTable from 'simple-json-table';
import _ from 'lodash';
import { omit, queryValue } from './object';
import { getLabel, getModelRelatedValue } from './config';
import Constants from '../constants';
import './string';

export const getModels = () => {
  return typeof window !== 'undefined' && window.sailsModels ?
  window.sailsModels : {};
};

export const getModel = (name) => {
  const models = getModels();
  return models[name] ? models[name] : {};
};


export const inCreationHiddenFields = (field) => {
  return [
    'createdAt',
    'updatedAt',
    'id',
  ].indexOf(field) > -1;
};

export const inUpdateHiddenFields = (field) => {
  return [
    'createdAt',
    'updatedAt',
    'id',
  ].indexOf(field) > -1;
};

const keysWeight = {
  id: -1,
  createdAt: 1,
  updatedAt: 2,
};
const weight = k => keysWeight[k] || 0;

const isADateDefaultField = (field) => {
  return [
    'createdAt',
    'updatedAt',
  ].indexOf(field) > -1;
};

export const keysSorter = (a, b) => weight(a) - weight(b);

export const getType = (model, field) => {
  if (isADateDefaultField(field)) return 'date';
  return model[field].type;
};

export const getFieldValueTemplate = (modelName, field) => {
  return getModelRelatedValue(
    `${modelName}.fields.${field}.valueTemplate`,
    null
  );
};

export const getModelValueTemplate = (modelName) => {
  return getModelRelatedValue(
    `${modelName}.valueTemplate`,
    null
  );
};

const isHTML = text => /<(\/)?\w+\s*>/.test(text);
const asHTML = (text) => {
  const value = { __html: text };
  return (
    <div dangerouslySetInnerHTML={value} />
  );
};

export const getModelValue = (modelName, item) => {
  const tpl = getModelValueTemplate(modelName);
  if (tpl) {
    const compiler = template(tpl);
    return compiler({ [modelName]: item, _ });
  }
};

export const valueResolver = (model, field, modelName) => (item) => {
  const tpl = getFieldValueTemplate(modelName, field);
  const data = item[field];

  if (tpl && data) {
    const compiler = template(tpl);
    const compiled = compiler({ [field]: data, _ });
    return isHTML(compiled) ? asHTML(compiled) : compiled;
  }
  
  if (isADateDefaultField(field)) {
    return new Date(+item[field]).toLocaleString();
  }

  if (model[field].type === 'boolean') {
    return item[field] ? 'true' : 'false';
  }

  if (model[field].model || model[field].collection) {
    return (
      <Popup
        trigger={<Button icon>{Constants.BUTTONS.SEE_DETAILS}</Button>}
        on='click'
      >
        <JSONTable
          source={omit(item[field], ['createdAt', 'updatedAt'])}
        />
      </Popup>
    ); 
  }

  return item[field];
};

export const getFieldLabel = (modelName, field) => {
  return getModelRelatedValue(
    `${modelName}.fields.${field}.label`,
    field.separateCamel().asTitle()
  );
};

export const getFieldRenderer = (modelName, field) => {
  return getModelRelatedValue(
    `${modelName}.fields.${field}.renderer`,
    field.separateCamel().asTitle()
  );
};

const CRUD_MODELS = [
  'crudaction',
  'crudresource',
  'crudright',
  'crudgroup',
  'cruduser',
];

export const NON_CRUD_MODELS_FILTER = (model) => {
  return CRUD_MODELS.indexOf(model) < 0
};

export const CRUD_MODELS_FILTER = (model) => {
  return CRUD_MODELS.indexOf(model) >= 0;
};

export default {
  getModel,
  getModels,
  inCreationHiddenFields,
  inUpdateHiddenFields,
  keysSorter,
  getType,
  valueResolver,
  getFieldLabel,
  getModelValue,
  getFieldRenderer,
  NON_CRUD_MODELS_FILTER,
  CRUD_MODELS_FILTER,
};
