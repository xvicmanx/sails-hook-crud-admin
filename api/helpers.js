const AuthService = require('./admin-services/AuthService');
const Errors  = require('./errors');


const MODELS_FILTER = (m) => {
  return m.indexOf('crudgroup_') < 0
};

const getDefinitions = (models) => {
  return Object.keys(models)
    .filter(MODELS_FILTER)
    .reduce(function (result, k) {
      result[k] = models[k].attributes;
      return result;
    }, {});
};

const populate = async (resultPromise, modelName, sails) => {
  const definition = getDefinitions(sails.models)[modelName];
  const fieldsToPopulate = Object.keys(definition).filter(k => {
      return definition[k].model || definition[k].collection;
    }).map(k => k);
  let result = resultPromise;
  fieldsToPopulate.forEach(field => {
    result = result.populate(field);
  })
  return result;
};

const queryValue = (source, query = '', defaultValue = null) => {
  const value = query.split('.').reduce((result, key) => (result && result[key] ? result[key] : null), source);
  return value || defaultValue;
};

const verifyToken = async (req, res) => {
  const token = req.headers['jwt-token'];
  const result = await AuthService.verifyToken(token);
  if (!result.success) {
    res.status(403)
      .send(Errors.tokenAuthError());
  }
  return result;
};

const verifyAccess = ({ action, resource }) => async (req, res) => {
  const verification = await verifyToken(req, res);
  const { data } = verification;
  const hasRightsAccess = AuthService.hasAccess({
    action,
    resource,
    rights: data && data.user &&
      data.user.rights || [],
  });
  return verification.success && hasRightsAccess;
};

module.exports = {
  verifyAccess,
  verifyToken,
  populate,
  getDefinitions,
  queryValue,
};
