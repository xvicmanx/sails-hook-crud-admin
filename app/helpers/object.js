export const queryValue = (source, query = '', defaultValue = null) => {
  const value = query.split('.').reduce((result, key) => (result && result[key] ? result[key] : null), source);
  return value || defaultValue;
};

export const anyPropLike = (props, value) => (obj) => {
  if (!props || props.length <= 0) return true;
  return props.reduce((acc, x) => {
    const target = (`${queryValue(obj, x, '')}`).toLowerCase();
    const needle = (`${value}`).toLowerCase();
    return acc || target.indexOf(needle) >= 0;
  }, false);
};

export const omit = (source, props) => {
  if (!source) return source;

  if (Array.isArray(source)) {
    const result = JSON.parse(JSON.stringify(source));
    result.forEach((element, index) => {
      result[index] = omit(result[index], props);
    });
    return result;
  }

  return Object.keys(source)
    .filter(k => props.indexOf(k) < 0)
    .reduce((acc, key) => {
      const result = acc;
      result[key] = source[key];
      return result;
    }, {});
};

export default {
  queryValue,
  anyPropLike,
  omit,
};
