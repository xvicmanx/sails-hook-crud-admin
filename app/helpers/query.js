const mapCondition = (x) => {
  switch (x) {
    case 'CONTAINS': return 'contains';
    case 'ENDS_WITH': return 'endsWith';
    case 'BEGINS_WITH': return 'startsWith';
    case 'IS': return 'is';
    case 'EQUALS_TO': return 'equals';
    case 'IS_NOT_EQUALS_TO': return '!=';
    case 'DOES_NOT_BEGIN_WITH': return '!=';
    case 'DOES_NOT_END_WITH': return '!=';
    case 'DOES_NOT_CONTAIN': return '!=';
    case 'LESS_THAN': return '<';
    case 'IS_NOT_LESS_THAN': '>=';
    case 'GREATER_THAN': return '>';
    case 'IS_NOT_GREATER_THAN': return '<=';
    case 'LESS_OR_EQUALS_THAN': return '<=';
    case 'IS_NOT_LESS_OR_EQUALS_THAN': return '>';
    case 'IS_NOT_LESS_OR_EQUALS_THAN': return '>';
    case 'GREATER_OR_EQUALS_THAN': return '>=';
    case 'IS_NOT_GREATER_OR_EQUALS_THAN': return '<';
    default: return 'contains';
  }
};

const getValue = (rule) => {
  if (
    rule.type === 'date'
    && typeof rule.value === 'string'
  ) {
    return new Date(rule.value).getTime();
  }
  return rule.value;
};

const buildQuery = rules => rules.reduce((query, rule) => {
  const condition = mapCondition(rule.condition);
  const value = getValue(rule);

  if (condition === 'equals') {
    query[rule.field] = value;
  } else if (condition === 'is') {
    query[rule.field] = value;
  } else {
    query[rule.field] = {
      [condition]: value,
    };
  }

  return query;
}, {});

module.exports = {
  buildQuery,
};
