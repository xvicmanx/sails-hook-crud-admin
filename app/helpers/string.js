/* eslint-disable */
String.prototype.asTitle = function separateCamel () {
  return this.charAt(0).toUpperCase() + this.slice(1);
};

String.prototype.separateCamel = function separateCamel () {
  return this.replace(/([a-z])([A-Z])/g, '$1 $2');
};

export default {};
