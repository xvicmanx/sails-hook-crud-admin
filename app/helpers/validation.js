export const validateModelRequiredValues = (model) => (values) => {
  const errors = {};
  const requiredFields = Object.keys(model)
    .filter((k) => model[k].required);
  requiredFields.forEach((k) => {
    if (!values[k]) {
      errors[k] = `Please, provide ${k}.`;
    }
  });
  return errors;
};

export const validateModelDeletion = (values) => {
  const errors = {};
  if (!values.id) {
    errors.id = 'Please, provide id';
  }
  return errors;
};

export default {
  validateModelRequiredValues,
  validateModelDeletion,
};
