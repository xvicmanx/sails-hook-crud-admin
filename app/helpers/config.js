import { queryValue } from './object';

export const getConfig = () => {
  return typeof window !== 'undefined' && window.crudAdminConfig ?
  window.crudAdminConfig : {};
};

export const getLabel = (prop, defaultValue) => {
  const config = getConfig();
  return queryValue(
    config,
    `general.labels.${prop}`,
    defaultValue
  );
};

export const getModelRelatedValue = (query, defaultValue) => {
  const config = getConfig();
  return queryValue(
    config,
    `models.${query}`,
    defaultValue
  );
};

export const getButtonText = (button, defaultValue) => {
  const config = getConfig();
  return queryValue(
    config,
    `general.buttons.${button}`,
    defaultValue
  );
};

export default {
  getConfig,
  getLabel,
  getButtonText,
  getModelRelatedValue,
};
