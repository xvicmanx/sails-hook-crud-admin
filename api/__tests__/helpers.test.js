import helpers from '../helpers';

describe('api::helpers', () => {
  describe('disableLog', () => {
    test('it disables sails log as expected', () => {
      const sails = {
        log: {
          verbose: 'test-verbose',
          error: 'test-error',
          silly: 'test-silly',
          info: 'test-info',
        },
      };
      helpers.disableLog(sails);
      expect(sails).toMatchSnapshot();
    });
  });
});
