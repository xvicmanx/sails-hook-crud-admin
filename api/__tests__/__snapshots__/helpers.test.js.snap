// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api::helpers disableLog it disables sails log as expected 1`] = `
Object {
  "log": [Function],
}
`;
